Replace `#[rustc_do_not_implement_via_object]` with `#[rustc_dyn_incompatible_trait]`

zachs18 opened the pull request:
Background: `#[rustc_do_not_implement_via_object]` on a trait currently still allows `dyn Trait` to exist (if the trait is otherwise dyn-compatible), it just means that `dyn Trait` does not automatically implement `Trait` via the normal object candidate. For some traits, this means that `dyn Trait` does not implement `Trait` at all (e.g. `Unsize` and `Tuple`). For some traits, this means that `dyn Trait` implements `Trait`, but with different associated types (e.g. `Pointee`, `DiscriminantKind`). Both of these cases can can cause issues with codegen , as seen in https://github.com/rust-lang/rust/issues/148089 (and https://github.com/rust-lang/rust/issues/148089#issuecomment-3447803823 ), because codegen assumes that if `dyn Trait` does not implement `Trait` (including if `dyn Trait<Assoc = T>` does not implement `Trait` with `Assoc == T`), then `dyn Trait` cannot be constructed, so vtable accesses on `dyn Trait` are unreachable, but this is not the case if `dyn Trait` has multiple supertraits: one which is `#[rustc_do_not_implement_via_object]`, and one which we are doing the vtable access to call a method from.

This PR replaces `#[rustc_do_not_implement_via_object]` with `#[rustc_dyn_incompatible_trait]`, which makes the marked trait dyn-incompatible, making `dyn Trait` not well-formed, instead of it being well-formed but not implementing `Trait`. This resolves rust-lang/rust#148089 by making it not compile.

May fix rust-lang/rust#148615 

The traits that are currently marked `#[rustc_do_not_implement_via_object]` are: `Sized`, `MetaSized`, `PointeeSized`, `TransmuteFrom`, `Unsize`, `BikeshedGuaranteedNoDrop`, `DiscriminantKind`, `Destruct`, `Tuple`, `FnPtr`, `Pointee`. Of these:
* `Sized` and `FnPtr` are already not dyn-compatible (`FnPtr: Copy`, which implies `Sized`)
* `MetaSized`
    * Removed `#[rustc_do_not_implement_via_object]`. Still dyn-compatible after this change. (Has a special-case in the trait solvers to ignore the object candidate for `dyn MetaSized`, since it `dyn MetaSized: MetSized` comes from the sized candidate that all `dyn Trait` get.)
* `PointeeSized`
    * Removed `#[rustc_do_not_implement_via_object]`. It doesn't seem to have been doing anything anyway ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=a395626c8bef791b87a2d371777b7841)), since `PointeeSized` is removed before trait solving(?).
* `Pointee`, `DiscriminantKind`, `Unsize`, and `Tuple` being dyn-compatible without having `dyn Trait: Trait` (with same assoc tys) can be observed to cause codegen issues (https://github.com/rust-lang/rust/issues/148089) so should be made dyn-incompatible
* `Destruct`, `TransmuteFrom`, and `BikeshedGuaranteedNoDrop` I'm not sure if would be useful as object types, but they can be relaxed to being dyn-compatible later if it is determined they should be.

-----

<details> <summary> resolved </summary>

Questions before merge:

1. `dyn MetaSized: MetaSized` having both `SizedCandidate` and `ObjectCandidate`
    1. I'm not sure if the checks in compiler/rustc_trait_selection/src/traits/project.rs and compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs were "load-bearing" for `MetaSized` (which is the only trait that was previously `#[rustc_do_not_implement_via_object]` that is still dyn-compatible after this change). Is it fine to just remove them? Removing them (as I did in the second commit) doesn't change any UI test results.
    3. IIUC, `dyn MetaSized` could get its `MetaSized` implementation in two ways: the object candidate (the normal `dyn Trait: Trait`) that was supressed by `#[rustc_do_not_implement_via_object]`, and the `SizedCandidate` (that all `dyn Trait` get for `dyn Trait: MetaSized`). Given that `MetaSized` has no associated types or methods, is it fine that these both exist now? Or is it better to only have the `SizedCandidate` and leave these checks in (i.e. drop the second commit, and remove the FIXMEs)?
    4. Resolved: the trait solvers special-case `dyn MetaSized` to ignore the object candidate in preference to the sizedness candidate (technically the check is for any `is_sizedness_trait`, but only `MetaSized` gets this far (`Sized` is inherently dyn-incompatible, and `dyn PointeeSized` is ill-formed for other reasons)
4. Diagnostics improvements?
    1. The diagnostics are kinda bad. If you have a `trait Foo: Pointee {}`, you now get a note that reads like *Foo* "opted out of dyn-compatbility", when really `Pointee` did that.
    2. Resolved: can be improved later
  
  <details> <summary>diagnostic example</summary>
  
```rs
#![feature(ptr_metadata)]

trait C: std::ptr::Pointee {}

fn main() {
    let y: &dyn C;
}
```
  
```rs
error[E0038]: the trait `C` is not dyn compatible
  --> c.rs:6:17
   |
 6 |     let y: &dyn C;
   |                 ^ `C` is not dyn compatible
   |
note: for a trait to be dyn compatible it needs to allow building a vtable
      for more information, visit <https://doc.rust-lang.org/reference/items/traits.html#dyn-compatibility>
  --> /home/zachary/opt_mount/zachary/Programming/rust-compiler-2/library/core/src/ptr/metadata.rs:57:1
   |
57 | #[rustc_dyn_incompatible_trait]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ...because it opted out of dyn-compatbility
   |
  ::: c.rs:3:7
   |
 3 | trait C: std::ptr::Pointee {}
   |       - this trait is not dyn compatible...

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0038`.
  ```
  
  </details> </details>
  
  Still investigating "3. `compiler/rustc_hir/src/attrs/encode_cross_crate.rs`: Should `DynIncompatibleTrait` attribute be encoded cross crate?"

zachs18 committed:
Replace #[rustc_do_not_implement_via_object] with #[rustc_dyn_incompatible_trait], which makes the marked trait dyn-incompatible.

Removes the attribute from `MetaSized` and `PointeeSized`, with a special case in the trait solvers for `MetaSized`.

`dyn MetaSized` is a perfectly cromulent type, and seems to only have had #[rustc_do_not_implement_via_object] so the builtin object
candidate does not overlap with the builtin MetaSized impl that all `dyn` types get.
Resolves this with a special case by checking `is_sizedness_trait` where the trait solvers previously checked `implement_via_object`.

`dyn PointeeSized` alone is rejected for other reasons (since `dyn PointeeSized` is considered to have no principal trait because `PointeeSized`
is removed at an earlier stage of the compiler), but `(dyn PointeeSized + Send)` is valid and equivalent to `dyn Send`.

Add suggestions from code review

Update compiler/rustc_trait_selection/src/traits/dyn_compatibility.rs and tests

Co-authored-by: lcnr <rust@lcnr.de>

rust-log-analyzer commented:
The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19162422666/job/54775426027) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/54775426027) [(plain)](https://github.com/rust-lang/rust/commit/7172834e1dcb67a94e0fd44cc2d80f489606c027/checks/54775426027/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
     ╭▸ compiler/rustc_feature/src/builtin_attrs.rs:1306:67
     │
1306 │         even if it otherwise satisfies the requirements to be dyn-comaptible."
     ╰╴                                                                  ━━━━━━━━━━
rerun tidy with `--extra-checks=spellcheck --bless` to fix typos
tidy [extra_checks]: checks with external tool 'typos' failed
tidy [extra_checks]: FAIL
tidy: The following check failed: extra_checks
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy /checkout /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /checkout/obj/build 4 /node/bin/npm --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1549:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1279:29

Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:02:30
  local time: Fri Nov  7 08:25:42 UTC 2025
  network time: Fri, 07 Nov 2025 08:25:42 GMT
##[error]Process completed with exit code 1.
```

</details>

theemathas commented:
Making `Destruct` be dyn-incompatible seems iffy to me...

bors commented:
:umbrella: The latest upstream changes (presumably #148753) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
r? types probably is a better assignment

zachs18 commented:
I added some background information about why #148089 happens and how this fixes it (by making it not compile), and a list of traits that will be affected by this change.

zachs18 commented:
```rs
#[rustc_do_not_implement_via_object]
trait NonObjectBase {}

trait ObjectBase {}

trait Foo: NonObjectBase + ObjectBase {}
```
> would an alternative way to fix this be to also disable the builtin trait object impl for traits with a `#[rustc_do_not_implement_via_object]` super trait?

So, in the example above, we would have that `dyn Foo` now doesn't implement `ObjectBase`? (it already doesn't implement `Foo` or `NonObjectBase`.) That might work, though I think we would also need to change some stuff about dyn upcasting (i.e. disallow dyn upcasting from `dyn Trait` if it doesn't implement `Trait`, as the codegen issues are reproducible with dyn upcasting as well ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=11fcd46d2eb178ba6350bb842cf7bd94)))[^1]



[^1]: We might also have codegen issues if `dyn Trait` has a manual impl of `Trait`, and then has subtraits, I'm not sure. But there is no case where that is possible at the moment since all `#[rustc_do_not_implement_via_object]` traits are also `#[rustc_deny_explicit_impl]`, so it's probably not important. In fact, it seems like the only trait that used to be `#[rustc_do_not_implement_via_object]` and allow user impls, `PointerLike`, [was the reason the two attributes were separated to begin with](https://github.com/rust-lang/rust/pull/134573), but [has since been removed](https://github.com/rust-lang/rust/pull/143308).

lcnr commented:
Hmmm, this change feels somewhat iffy to me and I can't quite put my finger on why. I feel like ideally I would spend a few hours to think about all of this myself to get a clear understanding.

As an alternative, would you be up to meet in sync to talk this through?

zachs18 commented:
I've rebased, and added comments explaining the special-case for `dyn MetaSized` to ignore the object candidate and only have the sizedness candidate, instead of having both.

I'm still looking in to "should it be encoded cross crate", but I'm leaning towards no since `#[rustc_do_not_implement_via_object]` was no, so that's what I have update it to now.

lcnr reviewed (commented):
r=me after nits. Is this in any way user-facing? I don't think it is. All the `#[rustc_do_not_implement_via_object]` traits are stable I think

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148637/22c74ba91873dd013479f86eac3e9ea10593bff9..10328b3e0c4cf30fc36175fcfa029f7c1e78d52d)*

zachs18 commented on tests/ui/traits/object/sizedness/dyn-metasized.rs:
```diff
@@ -0,0 +1,23 @@
+//@ run-pass
```
We have `traits/object`? I moved them to `traits/object/sizedness` (new subdir), does that sound good?

lcnr reviewed (commented):
r=me after final nit

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/148637/d940e56841ddcc05671ead99290e35ff2e98369f..1774dd89a574ecf758b4fd59d7dc3e006502994c)*
